### DISCUSSION TITLE
Baggage Claim Driver for EKS-based deployment

### DISCUSSION THREAD

**Discussion by matt-zego** [concourse/concourse#6384] [Help & Support]
We are deploying Concourse on Kubernetes/EKS, and we're running into issues with high I/O and CPU wait times on the nodes.  We've attributed these high times to cp -rf commands that are copying Concourse volumes from a "live" directory to an "init" directory.  We are using the naive baggage claim driver, and we'd like to change it but I'm not sure how to make it work.  The Amazon-provided EKS nodes do not have support btrfs in the kernel, so that leaves overlay as our only convenient option.  Today we attempted to launch a worker with the overlay driver and it started up fine, but it could not create volumes when it came time to execute:
```
{"timestamp":"2020-12-30T21:54:40.546977784Z","level":"error","source":"baggageclaim","message":"baggageclaim.api.volume-server.create-volume-async.create-volume.failed-to-materialize-strategy","data":{"error":"invalid argument","handle":"e85d6679-8e24-4d2f-58ce-a24e07ffe961","session":"3.1.2977.1"}}
{"timestamp":"2020-12-30T21:54:40.547005122Z","level":"error","source":"baggageclaim","message":"baggageclaim.api.volume-server.create-volume-async.failed-to-create","data":{"error":"invalid argument","handle":"e85d6679-8e24-4d2f-58ce-a24e07ffe961","privileged":true,"session":"3.1.2977","strategy":{"type":"cow","volume":"adcbc4bf-3721-4c46-54cb-0cca67b61b17"}}}
```
I logged into the concourse worker and observed the concourse work dir (off root) is already an overlay volume.  I'm not sure if that's significant:
```
root@concourse-worker-689c7b4f74-24lxx:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay         300G  100G  201G  34% /
tmpfs            64M     0   64M   0% /dev
tmpfs            16G     0   16G   0% /sys/fs/cgroup
/dev/nvme0n1p1  300G  100G  201G  34% /etc/hosts
tmpfs            16G  8.0K   16G   1% /concourse-keys
shm              64M     0   64M   0% /dev/shm
tmpfs            16G   12K   16G   1% /run/secrets/kubernetes.io/serviceaccount
```
The node supports overlay and docker is configured to use overlay2 backed by xfs, as shown by docker info:
```
Server:
 Containers: 18
  Running: 16
  Paused: 0
  Stopped: 2
 Images: 7
 Server Version: 19.03.6-ce
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
```
Any suggestions would be appreciated.  Naive is killing us, performance wise, but I can't seem to make anything else work.

**Comment by taylorsilva** [comment]
Are you able to manually create overlay mounts on the worker? On the worker can you try running this command; you'll need to create all four dir's first:

```
mount -t overlay overlay -o lowerdir=/lower,upperdir=/upper,workdir=/work /merged
```

**Reply by matt-zego** [reply]
Thank you for the response!
I get:
`wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error`

I believe the issue is that the worker is attempting to create an overlay file system volume on a volume that is already an overlay file system.  I did some more poking around and found a commit in the linux kernel that addresses stacked overlay file systems, outlined here:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1c8a47df36d72ace8cf78eb6c228aa0f8027d3c2 
and 
https://github.com/rkt/rkt/issues/1537

I'm not 100% sure that applies.

**Reply by taylorsilva** [reply]
Sadly I'm not too sure either. Not my area of expertise as well. Could you mount a different filesystem instead of overlay?

**Reply by matt-zego** [reply]
Thank you for your help.  We may be able to set up btrfs on the nodes and use that, or build a custom AMI, but I was hoping to use what's OOTB for EKS.
